pkg/antctl/command_definition.go

```diff
@@ -0,0 +1,425 @@
+		Short: cd.Short,
+		Long:  cd.Long,
+	}
+	renderDescription(cmd, isAgent)
+
+	cd.applyFlagsToCommand(cmd)
+	var keyArgOption *argOption
+	argOpts := cd.argOptions()
+	for i := range argOpts {
+		a := argOpts[i]
+		if a.key {
+			cmd.Use += fmt.Sprintf(" [%s]", a.name)
+			cmd.Long += "\n\nArgs:\n" + fmt.Sprintf("  %s\t%s", a.name, a.usage)
+			// save the key arg option
+			keyArgOption = a
+		}
+	}
+
+	if groupCommand, ok := groupCommands[cd.CommandGroup]; ok {
+		groupCommand.AddCommand(cmd)
+	} else {
+		root.AddCommand(cmd)
+	}
+	cd.applyExampleToCommand(cmd, keyArgOption)
+
+	// Set key arg length validator to the command.
+	if keyArgOption != nil {
+		cmd.Args = cobra.MaximumNArgs(1)
+	} else {
+		cmd.Args = cobra.NoArgs
+	}
+
+	cmd.RunE = cd.newCommandRunE(keyArgOption, client)
+}
+
+// validate checks if the commandDefinition is valid.
+func (cd *commandDefinition) validate() []error {
+	var errs []error
+	if len(cd.Use) == 0 {
+		errs = append(errs, fmt.Errorf("the command does not have name"))
+	}
+	if cd.TransformedResponse == nil {
+		errs = append(errs, fmt.Errorf("%s: command does not define output struct", cd.Use))
+	}
+	if !cd.Agent && !cd.Controller {
+		errs = append(errs, fmt.Errorf("%s: command does not define any supported component", cd.Use))
+	}
+	if cd.HandlerFactory == nil && len(cd.API) == 0 {
+		errs = append(errs, fmt.Errorf("%s: no handler or API specified", cd.Use))
+	}
+	if len(cd.API) != 0 && cd.Agent {
+		errs = append(errs, fmt.Errorf("%s: commands for agent do not allow to request API directly", cd.Use))
+	}
+	if cd.HandlerFactory != nil && cd.GroupVersion == nil {
+		errs = append(errs, fmt.Errorf("%s: must provide the group version of customize handler", cd.Use))
+	}
+	// Only one key arg is allowed.
+	var hasKey bool
+	for _, arg := range cd.argOptions() {
+		if arg.key && hasKey {
+			errs = append(errs, fmt.Errorf("%s: has more than one key field", cd.Use))
+			break
+		} else if arg.key {
+			hasKey = true
+		}
+	}
+	return errs
+}
+
+// argOptions returns the list of arguments that could be used in a commandDefinition.
+func (cd *commandDefinition) argOptions() []*argOption {
+	var ret []*argOption
+	if cd.TransformedResponse == nil {
+		return ret
+	}
+	t := reflect.Indirect(reflect.ValueOf(cd.TransformedResponse)).Type()
+
+	for i := 0; i < t.NumField(); i++ {
+		f := t.Field(i)
+		argOpt := &argOption{fieldName: f.Name}
+
+		tags, err := structtag.Parse(string(f.Tag))
+		if err != nil { // Broken cli tags, skip this field
+			continue
+		}
+
+		jsonTag, err := tags.Get("json")
+		if err != nil {
+			argOpt.name = strings.ToLower(f.Name)
+		} else {
+			argOpt.name = jsonTag.Name
+		}
+
+		cliTag, err := tags.Get(tagKey)
+		if err == nil {
+			argOpt.key = cliTag.Name == tagOptionKeyArg
+			argOpt.usage = strings.Join(cliTag.Options, ", ")
+		}
+
+		ret = append(ret, argOpt)
+	}
+	return ret
+}
+
+// decode parses the data in reader to the TransformedResponse. If the single is false,
```

Comment:
I suggest the following:
> decode parses the data in reader and converts it to one or more TransformedResponse objects. If single is false, the return type is []TransformedResponse. Otherwise, the return type is TransformedResponse.

Reply:
Updated.